builtin/bisect--helper.c

```diff
@@ -148,16 +170,81 @@ static int bisect_reset(const char *commit)
 	return bisect_clean_state();
 }
 
+static void log_commit(FILE *fp, char *fmt, const char *state,
+		       struct commit *commit)
+{
+	struct pretty_print_context pp = {0};
+	struct strbuf commit_msg = STRBUF_INIT;
+	char *label = xstrfmt(fmt, state);
+
+	format_commit_message(commit, "%s", &commit_msg, &pp);
+
+	fprintf(fp, "# %s: [%s] %s\n", label, oid_to_hex(&commit->object.oid),
+		commit_msg.buf);
+
+	strbuf_release(&commit_msg);
+	free(label);
+}
+
+static int bisect_write(const char *state, const char *rev,
+			const struct bisect_terms *terms, int nolog)
+{
+	struct strbuf tag = STRBUF_INIT;
+	struct object_id oid;
+	struct commit *commit;
+	FILE *fp = NULL;
+	int res = 0;
+
+	if (!strcmp(state, terms->term_bad)) {
+		strbuf_addf(&tag, "refs/bisect/%s", state);
+	} else if (one_of(state, terms->term_good, "skip", NULL)) {
+		strbuf_addf(&tag, "refs/bisect/%s-%s", state, rev);
+	} else {
+		res = error(_("Bad bisect_write argument: %s"), state);
+		goto finish;
+	}
+
+	if (get_oid(rev, &oid)) {
+		res = error(_("couldn't get the oid of the rev '%s'"), rev);
+		goto finish;
+	}
+
+	if (update_ref(NULL, tag.buf, &oid, NULL, 0,
+		       UPDATE_REFS_MSG_ON_ERR)) {
+		res = -1;
+		goto finish;
+	}
+
+	fp = fopen(git_path_bisect_log(), "a");
+	if (!fp) {
+		res = error_errno(_("couldn't open the file '%s'"), git_path_bisect_log());
+		goto finish;
+	}
+
+	commit = lookup_commit_reference(the_repository, &oid);
+	log_commit(fp, "%s", state, commit);
+
+	if (!nolog)
+		fprintf(fp, "git bisect %s %s\n", state, rev);
+
+finish:
+	if (fp)
+		fclose(fp);
```

Comment:
I think you can move this lines above the `finish:` label and remove the `if (fp)` line.

Reply:
But isn't that future proofing the code? You are suggesting to make it harder to improve the code because then everybody has to check carefully whether `goto finish` is enough to clean up resources or not.

Reply:
I think it is a question of preferences, policies and experience. Always thinking about possibly extending the code later can lead to over-engineering and in the end code that is more difficult to understand than necessary. Though of course if the code is later extended it can make that easier. As predicting the future is not easy, it is also not easy to know what's better now.

I am ok to keep it as is if you prefer.

Reply:
It is funny that you try to suggest to make the code harder to follow (what resources are released by the code after the label, what resources are not? is a question you force every reader to ask themselves) under the guise of trying to prevent over-engineering...

Reply:
For me the result would be easier to follow, not harder, as there is one less "if". I think in this specific case it depends on the way each person reads code if it is easier or harder to follow. Some people prefer to have resources released as soon as possible other prefer when they are all released at the end of a function if possible.

Reply:
Well, then you are probably unfamiliar with the `finally` construct in languages supporting exceptions. Because that is pretty precisely what this label is about.